src/devices/bus/s100/vectordualmode.cpp

```diff
@@ -0,0 +1,315 @@
+			ready = false;
+			track0 = flop && !flop->trk00_r();
+			seek_complete = false;
+			loss_of_sync = false;
+		}
+
+		data = write_protect | (ready << 1) | (track0 << 2)
+		    | (write_fault << 3) | (seek_complete << 4) | (loss_of_sync << 5)
+			| 0xc0;
+	} else if (offset == 0xc1) { // status (1) port
+		bool floppy_disk_selected;
+		bool controller_busy = m_sector_timer->enabled();
+		bool motor_on; // FDD
+		bool type_of_hard_disk = true;
+		if (hdd_selected()) {
+			floppy_disk_selected = false;
+			motor_on = false;
+		} else {
+			floppy_disk_selected = true;
+			motor_on = m_motor_on_timer->enabled();
+		}
+		data = floppy_disk_selected | (controller_busy << 1) | (motor_on << 2)
+		    | (type_of_hard_disk << 3) | 0xf0;
+	} else if (offset == 0xc2) { // data port
+		data = m_ram[m_cmar++];
+		m_cmar &= 0x1ff;
+	} else if (offset == 0xc3) { // reset port
+		m_cmar = 0;
+		data = 0xff;
+	} else {
+		data = 0xff;
+	}
+	return data;
+}
+
+void s100_vector_dualmode_device::s100_sout_w(offs_t offset, uint8_t data)
+{
+	// TODO: check actual behavior when controller is busy
+	if (m_sector_timer->enabled()) {
+		return;
+	}
+	// 7200-1200-02-1 page 14 (1-8)
+	if (offset == 0xc0) { // control (0) port
+		m_drive = BIT(data, 0, 2);
+		const uint8_t head = BIT(data, 2, 3);
+		const bool step = BIT(data, 5);
+		const bool step_in = BIT(data, 6);
+		//uint8_t low_current = BIT(data, 7);
+
+		for (int i = 0; i < m_floppy.size(); i++) {
+			floppy_image_device* flop = m_floppy[m_drive]->get_device();
+			if (flop)
+				flop->mon_w(0);
+		}
+		// WR0| triggers U60, a 74LS123 with 100uF cap and 100k res
+		m_motor_on_timer->adjust(attotime::from_usec(2819600));
+
+		floppy_image_device* flop = m_floppy[m_drive]->get_device();
+		if (flop) {
+			flop->ss_w(head & 1);
+			// Software should not change other bits when pulsing step
+			flop->stp_w(!step);
+			flop->dir_w(!step_in);
+		}
+	} else if (offset == 0xc1) { // control (1) port
+		m_sector = BIT(data, 0, 5);
+		m_read = BIT(data, 5);
+	} else if (offset == 0xc2) { // data port
+		m_ram[m_cmar++] = data;
+		m_cmar &= 0x1ff;
+	} else if (offset == 0xc3) { // start port
+		floppy_image_device* flop = m_floppy[m_drive]->get_device();
+		if (!flop || flop->time_next_index().is_never())
+			return;
+		const attotime rot_time = attotime::from_msec(200);
+		attotime sector_time = flop->time_next_index() - machine().time() + (rot_time / 16) * m_sector;
+		if (sector_time > rot_time)
+			sector_time -= rot_time;
+		m_sector_timer->adjust(sector_time, SECTOR_START);
+	}
+}
+
+bool s100_vector_dualmode_device::get_next_bit(attotime &tm, const attotime &limit)
+{
+	int bit = m_pll.get_next_bit(tm, m_floppy[m_drive]->get_device(), limit);
+	if (bit < 0)
+		return false;
+	m_pending_byte <<= 1;
+	m_pending_byte |= bit;
+	m_pending_size++;
+	return true;
+}
+
+TIMER_CALLBACK_MEMBER(s100_vector_dualmode_device::sector_cb)
+{
+	switch (static_cast<sector_timer_state>(param)) {
+	case SECTOR_START:
```

Comment:
Just a nitpick, but since `sector_timer_state` is an unscoped enum, you don’t actually need the cast here at all, and even if you want to use it, a function syntax cast will work fine with less visual clutter (i.e. `switch (sector_timer_state(param)) {` is equivalent with less clutter).